openeo/result.py

```diff
@@ -0,0 +1,76 @@
+from pathlib import Path
+from typing import Union, Dict
+
+from openeo.rest import OpenEoClientException
+from openeo.util import ensure_dir
+
+
+class Result:
+    def __init__(self, job):
+        self.job = job
+
+    def download_file(self, target: Union[str, Path] = None) -> Path:
+        assets = self._download_get_assets()
+        if len(assets) != 1:
+            raise OpenEoClientException(
+                "Expected one result file to download, but got {c}: {u!r}".format(c=len(assets), u=assets))
+        filename, metadata = next(iter(assets.items()))
+        url = metadata["href"]
+
+        target = Path(target or Path.cwd())
+        if target.is_dir():
+            target = target / filename
+
+        self._download_url(url, target)
+        return target
+
+    def download_files(self, target: Union[str, Path] = None) -> Dict[Path, dict]:
+        target = Path(target or Path.cwd())
+        if target.exists() and not target.is_dir():
+            raise OpenEoClientException("The target argument must be a folder. Got {t!r}".format(t=str(target)))
+
+        assets = {target / f: m for (f, m) in self._download_get_assets().items()}
+        if len(assets) == 0:
+            raise OpenEoClientException("Expected at least one result file to download, but got 0.")
+
+        for path, metadata in assets.items():
+            self._download_url(metadata["href"], path)
+
+        return assets
+
+    def _download_url(self, url: str, path: Path):
+        ensure_dir(path.parent)
+        with path.open('wb') as handle:
+            # TODO: finetune download parameters/chunking?
+            response = self.job.connection.get(url, stream=True)
+            for block in response.iter_content(1024):
+                if not block:
+                    break
+                handle.write(block)
+        return path
+
+    def load_json(self) -> dict:
+        return self._get_response_object().json()
+
+    def load_bytes(self):
+        return self._get_response_object().content
+
+    def _get_response_object(self):
+        assets = self._download_get_assets()
+        if len(assets) != 1:
+            raise OpenEoClientException(
+                "Expected one result file to download, but got {c}: {u!r}".format(c=len(assets), u=assets))
+        filename, metadata = next(iter(assets.items()))
+        url = metadata["href"]
+
+        return self.job.connection.get(url, stream=True)
+
+    def _download_get_assets(self) -> Dict[str, dict]:
```

Comment:
minor: you can drop that `_download` prefix in the function name.
I used that under the `RESTJob` class to make clear that was about downloads, but here it is bit redundant